### PR TITLE
resolve full SEED ID of picks when reading NLLOC HYP files and nordic files, refactor the functionality for sh evt and hypodd pha files

### LINKED ISSUE #1587: Nonlinloc to quakeml - picks are not associated with any channels
Dear all

When I am trying to convert nonlinloc file to quakeml event file, phases from nlloc are not associated with chans in quakeml file, although nlloc file does have them associated.

original issue:
https://github.com/eqcorrscan/EQcorrscan/issues/69

nlloc file:
https://drive.google.com/open?id=0BzMcKedS7yadaFRjZlFaYUpOQkU

I am using the script as simple as:
```
import glob
import os
from obspy import Catalog,read_events

cat = Catalog()

path = 'nllocs/'
event_files = glob.glob(path + '*.hyp')

for i, event_file1 in enumerate(event_files):

    cat += read_events(event_file1)
    print "Reading", i+1, "of", len(event_files)
    
    for event in cat:
    
        event.write("events/"+str(event.origins[0].time)+".xml", format="QUAKEML")
```
Is the channel association something that should be in quakeml format? the website https://quake.ethz.ch/quakeml/ is not working and I am unable to find this info.

Thanks,
blaz

### PR DESCRIPTION
**Pull request by megies**
### What does this PR do?

Fixes #1587 

### Why was it initiated?  Any relevant Issues?

See #1587 

### PR Checklist
- [ ] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+ DOCS"
- [ ] If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:"
      (you can also add "ALL" to just simply run all tests across all modules)
- [ ] All tests still pass.
- [ ] Any new features or fixed regressions are be covered via new tests.
- [ ] Any new or changed features have are fully documented.
- [ ] Significant changes have been added to `CHANGELOG.txt` .
- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .

Might need some additional polishing/checking but mostly ready. CC @bvicic

### CODE REVIEW

**Comment by trichter** [comment]
Just stumbled upon this PR. I think it would be good to unify the interface of the different io routines, that have a need for this functionality:
nlloc hyp, sh evt, hypodd pha. Any other format?

**Comment by d-chambers** [comment]
The nordic format (used by seisan) has the same issue. I dont know if ObsPy's nordic parser does any re-association.

**Comment by calum-chamberlain** [comment]
The nordic parser does not do any reassociation.

CJ Chamberlain, out of office

________________________________

**Comment by megies** [comment]
@trichter feel free to continue work on this PR, dont really remember why i made it and then never merged it

**Comment by trichter** [comment]
I put some effort into this PR, feel free to comment or review.

**Comment by trichter** [comment]
+DOCS

**Comment by trichter** [comment]
I followed the first commit by megies and network and location code are set to None by default (in WaveformStreamID). When writing to QuakeML, the corresponding entries are not present in WaveformID. I do not know if this is valid QuakeML? I had problems reading the QuakeML with these missing codes with other software.

File: CHANGELOG.txt [hunk 1]
```diff
@@ -7,6 +7,11 @@ Changes:
      instrument response (see #2677)
  - obspy.clients.filesystem:
    * add get_waveforms_bulk() method to SDS client (see #2616, #2626)
+ - obspy.io
+   * add support to resove the SEED id of picks for nlloc hyp files and
```
Comment (megies) [inline CHANGELOG.txt]:
```suggestion
   * add support to resolve the SEED id of picks for nlloc hyp files and
```

File: obspy/io/nlloc/tests/data/nlloc_custom.qml [hunk 2]
```diff
@@ -157,63 +157,63 @@ STATISTICS ExpectX 4473.68 Y 5323.29 Z 4.59501  CovXX 0.0282621 XY -0.0053866 XZ
...
+        <waveformID channelCode="?" stationCode="UH3"></waveformID>
         <phaseHint>P</phaseHint>
       </pick>
       <pick publicID="smi:local/3a0bde89-d7e6-45ef-a08a-4b950720f7be">
         <time>
           <value>2010-05-27T16:56:27.100000Z</value>
           <uncertainty>0.06</uncertainty>
         </time>
-        <waveformID networkCode="" stationCode="UH3"></waveformID>
+        <waveformID channelCode="?" stationCode="UH3"></waveformID>
```
Comment (megies) [inline obspy/io/nlloc/tests/data/nlloc_custom.qml:168]: channel code `"?"`?

**Comment by megies** [comment]
@trichter net/sta are required, loc/cha optional:

```xml
391   <xs:complexType name="WaveformStreamID">
392     <xs:simpleContent>
393       <xs:extension base="bed:ResourceReference_optional">
394         <xs:attribute name="networkCode" use="required">
395 +-------  5 lines: <xs:simpleType>--------------------------------------------------------------------------------------------
400         </xs:attribute>
401         <xs:attribute name="stationCode" use="required">
402 +-------  5 lines: <xs:simpleType>--------------------------------------------------------------------------------------------
407         </xs:attribute>
408         <xs:attribute name="channelCode">
409 +-------  5 lines: <xs:simpleType>--------------------------------------------------------------------------------------------
414         </xs:attribute>
415         <xs:attribute name="locationCode">
416 +-------  5 lines: <xs:simpleType>--------------------------------------------------------------------------------------------
421         </xs:attribute>
422         <xs:anyAttribute namespace="##other" processContents="lax"/>
423       </xs:extension>
424     </xs:simpleContent>
425   </xs:complexType>
```

Without looking at the details, if net/sta are unknown setting them to an empty string before saving to QuakeML might be the only way to get a valid file.
I also just noticed that `WaveformStreamID` can be initialized without net/sta which will result in problems with QuakeML output. Maybe we should show a warning on setting `WaveformStreamID.network_code` (and station_code) with a `None` value.

File: obspy/io/nlloc/tests/data/nlloc_custom.qml (hunk 2 above)
Comment (trichter) [reply obspy/io/nlloc/tests/data/nlloc_custom.qml:168]: Yea, these where in [one of the nlloc hyp files](https://github.com/obspy/obspy/blob/master/obspy/io/nlloc/tests/data/nlloc_custom.hyp#L17). I don't know. What do you suggest? Maybe do not set the channel when it is `?`. What with channel codes like `HH?`? Additionally raise a warning?

**Comment by trichter** [comment]
> net/sta are required, loc/cha optional

Thanks for looking that up.

File: CHANGELOG.txt (hunk 1 above)
Comment (trichter) [reply CHANGELOG.txt]: This was already fixed. Did you review the latest version?

**Comment by trichter** [comment]
@megies I incorporated your comments. Maybe you can have another look. I am not an expert of nlloc file formats.

File: obspy/core/inventory/util.py [hunk 3]
```diff
@@ -1115,16 +1115,30 @@ def _add_resolve_seedid_doc(func):
     return func
 
 
+def _add_resolve_seedid_ph2comp_doc(func):
+    """
+    :param dict ph2comp: mapping of phases to components if format does not
+        specify the component or if the component ends with '?'. Set it to
+        `None` for no mapping of components. (default: {'P': 'Z', 'S': 'N'})
+    """
+    func.__doc__ = func.__doc__ + __doc__
```
Comment (QuLogic) [inline obspy/core/inventory/util.py:1124]: Check if `func.__doc__ is None`, or this will crash when run with optimization enabled.

Comment (trichter) [reply obspy/core/inventory/util.py:1124]: Thank you, I didn't know that doc strings were eliminated with optimization. Might also be applicable to other parts of the code base.

**Comment by trichter** [comment]
Can we add this to the new release? I will put some work into it and try to rebase onto master.

**Comment by ThomasLecocq** [comment]
> Can we add this to the new release? I will put some work into it and try to rebase onto master.

Hey @trichter sure, please rebase on master, force-push so we can see the CI color!

**Comment by trichter** [comment]
I think this is ready.

File: obspy/io/sh/tests/test_evt.py [hunk 4]
```diff
@@ -91,8 +91,9 @@ def test_populate_waveform_id(self):
         # read invenroty - FUR with channels, WET without channels
         inv = read_inventory().select(channel='HH?')
         inv[0][1].channels = []
-        picks = read_events(fname, inventory=inv, id_default='BLA.{}.11.DH{}',
-                            id_map={'UBR': 'BLB.{}.00.BH{}'})[0].picks
+        cat = read_events(fname, inventory=inv, id_default='BLA.{}.11.DH{}',
+                          id_map={'UBR': 'BLB.{}.00.BH{}'}, warn=False)
```
Comment (ThomasLecocq) [inline obspy/io/sh/tests/test_evt.py:95]: here passing `id_map` and not `seedid_map` is voluntary?

**Review by ThomasLecocq — APPROVED** [review]

File: obspy/io/sh/tests/test_evt.py (hunk 4 above)
Comment (trichter) [reply obspy/io/sh/tests/test_evt.py]: Thanks for spotting this. The use of `id_map` and `id_default` are non-intentional here. They are only supported for backwards compatibility.

**Comment by ThomasLecocq** [comment]
@trichter for next big merges, it's good to "squash & merge" so that the PR is a single large commit :-)

**Pull request merged**